.github/workflows/pytest.yml

```diff
@@ -12,18 +16,41 @@ jobs:
         python-version: ["3.10"]
         os: ['ubuntu-latest', 'windows-latest', 'macos-latest']
     steps:
-    - uses: actions/checkout@v2
+    - uses: actions/checkout@v3
     - name: Set up ${{ matrix.python-version }}
-      uses: actions/setup-python@v1
+      uses: actions/setup-python@v4
+      id: py
       with:
         python-version: ${{ matrix.python-version }}
-    - name: Install dependencies
+    - uses: actions/cache@v3
+      id: cache
+      with:
+        path: venv
+        key: ${{ runner.os }}-${{ steps.py.outputs.python-version }}-${{ hashFiles('requirements.txt') }}
```

Comment:
The way this works is, a key is computed from the `os-python_version-hash(requirements.txt)` , that key is looked up in the cache and it is either found (cache-hit = true) or not found (cache-hit = false) . 

In the case that it is found:
1. it copies the `venv` path from the path to the local folder
2. the pip install steps are skipped
3. we activate the virtual env and run pytest

In the case that it is not found (requirements change):
1. create empty venv, activate, then pip install requirements
2. run pytest in that environment
3. if the job is successful the venv folder is cached under the new key